rust/connlib/tunnel/src/peer.rs

```diff
@@ -578,56 +586,41 @@ impl TranslationState {
-            self.missing_responses_detected_at
-                .is_some_and(|missing_responses_detected_at| {
-                    now.duration_since(missing_responses_detected_at) >= Duration::from_secs(1)
-                });
+        // Note: we don't need to check that it's used here because the ack grace period already implies it
+        self.ack_grace_period_expired(now) && self.no_incoming_in_120s(now)
+    }
 
-        self.no_incoming_in_120s(now) && self.is_used(now) && ack_grace_period_expired
+    fn ack_grace_period_expired(&self, now: Instant) -> bool {
+        self.ack_grace_period_started_at
+            .is_some_and(|missing_responses_detected_at| {
+                now.duration_since(missing_responses_detected_at) >= Duration::from_secs(1)
+            })
     }
 
     fn no_incoming_in_120s(&self, now: Instant) -> bool {
         const CONNTRACK_UDP_STREAM_TIMEOUT: Duration = Duration::from_secs(120);
 
-        self.no_incoming_in_window(now - CONNTRACK_UDP_STREAM_TIMEOUT, now)
-    }
-
-    fn no_incoming_in_window(&self, start: Instant, finish: Instant) -> bool {
-        let Some(last_incoming) = self.last_incoming else {
-            return true;
-        };
-
-        debug_assert!(start <= finish);
-
-        let before_start = last_incoming < start;
-        let after_finish = last_incoming > finish;
-
-        before_start || after_finish
+        if let Some(last_incoming) = self.last_incoming {
+            now.duration_since(last_incoming) >= CONNTRACK_UDP_STREAM_TIMEOUT
+        } else {
+            now.duration_since(self.created_at) >= CONNTRACK_UDP_STREAM_TIMEOUT
+        }
     }
 
     fn on_incoming_traffic(&mut self, now: Instant) {
         self.last_incoming = Some(now);
-    }
-
-    fn missing_multiple_responses(&self, now: Instant) -> bool {
-        /// The maximum time we expect a packet to be acknowledged in.
-        const MAX_ACK_DELAY: Duration = Duration::from_secs(110);
-
-        let no_incoming_before_last_packet = self
-            .last_outgoing
-            .map(|last_outgoing| {
-                self.no_incoming_in_window(last_outgoing - MAX_ACK_DELAY, last_outgoing)
-            })
-            .unwrap_or(true); // No outgoing packet means we also don't have a response!
-        let no_incoming_now = self.no_incoming_in_window(now - MAX_ACK_DELAY, now);
-
-        no_incoming_now && no_incoming_before_last_packet
+        self.ack_grace_period_started_at = None;
     }
 
     fn on_outgoing_traffic(&mut self, now: Instant) {
-        if self.missing_multiple_responses(now) {
-            self.missing_responses_detected_at = Some(now);
+        // We need this because it means that if a packet arrives at some point less than 120s but more than 110s
+        // we still start the grace period so that the connection expires
+        let with_this_packet_the_connection_will_be_considered_used_when_it_expires =
+            self.no_incoming_in_120s(now + Self::USED_WINDOW);
```

Comment:
I don't understand why we set `now` to a value here that points into the future. There must be a better way to express this, no? We can never see packets from the future!

Reply:
This just means that the packet we are currently seeing will count as the "connection being used" when this expires.

If we only checked it for `now`, means that an outgoing packet at 110s of no incoming packet wouldn't trigger the expiration, because we would never start the ack grace period, therefore it won't have passed.

I think there might be a way of making this a bit clearer if we make `is_expired` mutable(we could rename it), so that if the connection is used it would trigger at that point the ack grace period. 

Or maybe we can just add a `handle_timeout` and `poll_timeout` here-- but that might be overcomplicating things a bit.

This would require changing a lot of the tests, to consider the extra 1 second delay after 120s.

I will be merging this for now, so we can start testing it, and we can consider this later on.

Reply:
also, just to clarify, it's not a packet from the future, it's more like

"With no additional incoming packet, this one will trigger a refresh"